Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -240,51 +339,68 @@ public struct HTTPHeaders: CustomStringConvertible {
-        buffer.write(string: key)
-        buffer.write(staticString: headerSeparator)
+    /// Checks if a header is present
+    ///
+    /// - parameters:
+    ///     - name: The name of the header
+    ///     - value: The value of the header or nil if the name should not be checked
+    //  - returns: `true` if a header with the name (and value) exists, `false` otherwise.
+    public func contains(name: String, value: String? = nil) -> Bool {
+        guard !self.headers.isEmpty else {
+            return false
+        }
 
-        var writerIndex = buffer.writerIndex
-        for (_, value) in values {
-            buffer.write(string: value)
-            writerIndex = buffer.writerIndex
-            buffer.write(staticString: ",")
+        let utf8 = name.utf8
+        let stringLength = utf8.count
+        for header in headers {
+            if stringLength == header.name.length && self.buffer.equalCaseInsensitiveASCII(seq: utf8, at: header.name) {
+                if let value = value {
+                    return value.count == header.value.length && self.buffer.getString(at: header.value.start, length: header.value.length) == value
+                }
+                return true
+            }
         }
-        // Discard last ,
-        buffer.moveWriterIndex(to: writerIndex)
-        buffer.write(staticString: crlf)
+        return false
     }
 
-    /// Used for HTTP headers that cannot be joined with commas, e.g. set-cookie.
-    private func writeSequentialHeaderValues(buffer: inout ByteBuffer, key: String, values: [(String, String)]) {
-        for (_, value) in values {
-            buffer.write(string: key)
-            buffer.write(staticString: headerSeparator)
-            buffer.write(string: value)
-            buffer.write(staticString: crlf)
+    /// Serializes this HTTP header block to bytes suitable for writing to the wire.
+    ///
+    /// - Parameter buffer: A buffer to write the serialized bytes into. Will increment
+    ///     the writer index of this buffer.
+    func write(into: inout ByteBuffer) {
+        if self.continuous {
+            // Declare an extra variable so we not affect the readerIndex of the buffer itself.
+            var buf = self.buffer
+            into.write(buffer: &buf)
+        } else {
+            // slow-path....
+            // TODO: This can still be improved to write as many continuous data as possible and just skip over stuff that was removed.
+            for header in headers {
+                var name = self.buffer.getSlice(at: header.name.start, length: header.name.length)!
+                into.write(buffer: &name)
+                into.write(staticString: headerSeparator)
+
+                var value = self.buffer.getSlice(at: header.value.start, length: header.value.length)!
+                into.write(buffer: &value)
+                into.write(staticString: crlf)
```

Comment:
Right now names and values are always contiguous, right? If they are, we can halve this out by writing just one slice:

```swift
var header = self.buffer.getSlice(at: header.name.start, length: header.name.length + header.value.length + 4)
into.write(buffer: &header)
```

Reply:
@Lukasa I think this will not work in all cases as we can't be sure it uses `: `... It could also just use `:` (which is legit as well).

Reply:
Why can't we be sure? Whatever we're sending out was either written there by us, or it was provided by http_parser, but in either case it did parse.

Reply:
Sure, but does that matter? The leading whitespace is explicitly optional. RFC 7230 says it right in the ABNF.

Reply:
> Each header field consists of a case-insensitive field name followed by a colon (":"), *optional* leading whitespace, the field value, and *optional* trailing whitespace.

(Emphasis mine)

Reply:
@Lukasa and thats exactly the problem ... 

```
self.buffer.getSlice(at: header.name.start, length: header.name.length + header.value.length + 4)
```

This will only work with `: `. Without the whitespace it would be `+3`.

Reply:
Oh, sure, but that's ok, just omit that part of the logic and do this:

```
let fieldLength = (header.value.start + header.value.length) - header.name.start
var header = self.buffer.getSlice(at: header.name.start, length: fieldLength)
into.write(buffer: &header)
into.write(staticString: "\r\n")
```

That also has the effect of stripping any OWS at the end of the field.

Reply:
@Lukasa good point.. fixing now